vm/src/builtins/pystr.rs

```diff
@@ -122,20 +129,71 @@ impl PyValue for PyStrIterator {
 }
 
 #[pyimpl(with(PyIter))]
-impl PyStrIterator {}
+impl PyStrIterator {
+    #[pymethod(magic)]
+    fn length_hint(&self) -> usize {
+        match self.status.load() {
+            Active => {
+                let pos = self.position.load(atomic::Ordering::SeqCst);
+                self.string.len().saturating_sub(pos)
+            }
+            Exhausted => 0,
+        }
+    }
+
+    #[pymethod(magic)]
+    fn setstate(&self, state: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
+        // When we're exhausted, just return.
+        if let Exhausted = self.status.load() {
+            return Ok(());
+        }
+        let pos = state
+            .payload::<PyInt>()
+            .ok_or_else(|| vm.new_type_error("an integer is required.".to_owned()))?;
+        let pos = std::cmp::min(
+            try_to_primitive(pos.as_bigint(), vm).unwrap_or(0),
+            self.string.len(),
+        );
+        self.position.store(pos, atomic::Ordering::SeqCst);
+        Ok(())
+    }
+
+    #[pymethod(magic)]
+    fn reduce(&self, vm: &VirtualMachine) -> PyResult {
+        let pos = if let Exhausted = self.status.load() {
+            None
+        } else {
+            Some(self.position.load(atomic::Ordering::Relaxed))
+        };
+        let iter = vm.get_attribute(vm.builtins.clone(), "iter")?;
+        let elems = match pos {
+            None => vec![iter, vm.ctx.new_tuple(vec![vm.ctx.new_str("")])],
+            Some(pos) => vec![
+                iter,
+                vm.ctx.new_tuple(vec![self.string.clone().into_object()]),
+                vm.ctx.new_int(pos),
+            ],
+        };
+        Ok(vm.ctx.new_tuple(elems))
```

Comment:
```suggestion
        let iter = vm.get_attribute(vm.builtins.clone(), "iter")?;
        Ok(match self.status.load() {
            Exhausted => vec![iter, vm.ctx.new_tuple(vec![vm.ctx.new_str("")])],
            Active => vec![
                iter,
                vm.ctx.new_tuple(vec![self.string.clone().into_object()]),
                vm.ctx.new_int(self.position.load(atomic::Ordering::Relaxed)),
            ],
        });
```

Reply:
It looks even better! I applied your suggestion. Thanks for reviewing.